scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/module/MnistMlp.scala

```diff
@@ -0,0 +1,158 @@
+    val act1 = Symbol.Activation(name = "relu1")(fc1)(Map("act_type" -> "relu"))
+    val fc2 = Symbol.FullyConnected(name = "fc2")(act1)(Map("num_hidden" -> 64))
+    val act2 = Symbol.Activation(name = "relu2")(fc2)(Map("act_type" -> "relu"))
+    val fc3 = Symbol.FullyConnected(name = "fc3")(act2)(Map("num_hidden" -> 10))
+    val softmax = Symbol.SoftmaxOutput(name = "softmax")(fc3)()
+    softmax
+  }
+
+  def runIntermediateLevelApi(train: DataIter, eval: DataIter, cmdLine: MnistMlp): Unit = {
+    // Intermediate-level API
+    val mod = new Module(getSymbol)
+    mod.bind(dataShapes = train.provideData, labelShapes = Some(train.provideLabel))
+    mod.initParams()
+    mod.initOptimizer(optimizer = new SGD(learningRate = 0.01f, momentum = 0.9f))
+
+    val metric = new Accuracy()
+
+    for (epoch <- 0 until cmdLine.numEpoch) {
+      while (train.hasNext) {
+        val batch = train.next()
+        mod.forward(batch)
+        mod.updateMetric(metric, batch.label)
+        mod.backward()
+        mod.update()
+      }
+
+      val (name, value) = metric.get
+      logger.info(s"epoch $epoch $name=$value")
+      metric.reset()
+      train.reset()
+    }
+
+    // High-level API
+    train.reset()
+  }
+
+  def runHighLevelApi(train: DataIter, test: DataIter, cmdLine: MnistMlp): Unit = {
+    val mod = new Module(getSymbol)
+    mod.fit(train, evalData = scala.Option(test), numEpoch = cmdLine.numEpoch)
+
+    // prediction iterator API
+    var iBatch = 0
+    test.reset()
+    while (test.hasNext) {
+      val batch = test.next()
+      val preds = mod.predict(batch)
+      val predLabel: Array[Int] = NDArray.argmax_channel(preds(0)).toArray.map(_.toInt)
+      val label = batch.label(0).toArray.map(_.toInt)
+      val acc = (predLabel zip label).map { case (py, y) =>
+        if (py == y) 1 else 0
+      }.sum / predLabel.length.toFloat
+      if (iBatch % 20 == 0) {
+        logger.info(s"Batch $iBatch acc: $acc")
+      }
+      iBatch += 1
+    }
+
+    // a dummy call just to test if the API works
+    mod.predict(test)
+
+    // perform prediction and calculate accuracy manually
+    val preds = mod.predictEveryBatch(test)
+    test.reset()
+    var accSum = 0.0f
+    var accCnt = 0
+    var i = 0
+    while (test.hasNext) {
+      val batch = test.next()
+      val predLabel: Array[Int] = NDArray.argmax_channel(preds(i)(0)).toArray.map(_.toInt)
+      val label = batch.label(0).toArray.map(_.toInt)
+      accSum += (predLabel zip label).map { case (py, y) =>
+        if (py == y) 1 else 0
+      }.sum
+      accCnt += predLabel.length
+      i += 1
+    }
+    logger.info(s"Validation Accuracy: {}", accSum / accCnt.toFloat)
+
+    // evaluate on validation set with a evaluation metric
+    val (name, value) = mod.score(test, new Accuracy).get
+    logger.info("Scored {} = {}", name, value)
+  }
+
+  def main(args: Array[String]): Unit = {
+    val inst = new MnistMlp
+    val parser: CmdLineParser = new CmdLineParser(inst)
```

Comment:
is it a necessary claim of the variable type? (I mean we may only need to explicitly claim the local variable's type when it is hard to be detected at the first sight)